### ISSUE TITLE
Membership plugin

### ISSUE THREAD

**Issue by pavlenex** [btcpayserver/btcpayserver#6447]
Placeholder, full feature document available on request. 
A plugin allowing server administrators to charge re-occuring payments/subscriptions to user registred on their instance.

**Comment by NicolasDorier** [comment]
Note to self
See what rockstardev did https://github.com/rockstardev/BTCPayServerPlugins.RockstarDev/tree/master/Plugins/BTCPayServer.RockstarDev.Plugins.Subscriptions
See what kukks did https://github.com/Kukks/BTCPayServerPlugins/blob/master/Plugins/BTCPayServer.Plugins.Subscriptions
See what tobses did https://github.com/TChukwuleta/BTCPayServerPlugins/tree/main/Plugins/BTCPayServer.Plugins.GhostPlugin

**Comment by NicolasDorier** [comment]
# Main Motivation

Enable ambassadors to monetize their BTCPay Instance by offering memberships to the users they onboard.

# Concepts

* A **Membership Program** is a store-level concept that defines what is being sold (for example, access to BTCPay Server).
* Each membership program contains different **Subscription Plans**, which represent the various ways of subscribing (e.g., 1 month, 3 months, 6 months, or 1 year).
* A **Subscription** is created when a customer chooses a plan. It ties a plan to a specific customer.
* A **Customer** is a store-level concept.

## Customers

The concept of a customer is useful in areas beyond subscriptions (for example, Pay Vendor plugin).  
In the future, we could add the ability to associate a customer with an invoice. This would allow features like filtering invoices by customer.

## Membership

The concept of membership can also be useful in contexts unrelated to granting access to a BTCPay Server instance.  
As such, a store may host more than one membership program. In BTCPay terminology, each membership program would be represented as an App.

# Flow

## Merchant: Creating a Membership Program

The merchant creates a new Membership Program app in BTCPay Server.  
They configure the membership settings, which include creating a form for customer sign-up.

Next, the merchant can add and configure each subscription plan:

* **Name / DisplayName** – e.g., “1-Month Membership”, “Quarterly Membership”.
* **Description** – marketing copy or internal notes about what’s included.
* **BillingInterval** – the unit of time between renewals (e.g., “month”, “year”).
* **IntervalCount** – a multiplier of the unit (e.g., 1 → monthly, 3 → quarterly).
* **PriceAmount** – the numeric cost per interval.
* **Currency** – e.g., “USD”, “EUR”.
* **TrialPeriodDays** – the number of free days before the first charge.
* **Status** – active, retired, or draft.

## Customer Onboarding

* The merchant displays a QR code from the app page to the customer.  
* The customer fills out a form configured by the merchant (typically Name, Email, Password, and Plan).  
* A customer record is created.  
* A subscription is created. The customer is redirected to the subscription page, where they can pay for their first billing period.

## Admin: Connecting a Membership Program to BTCPay Server Users (Server Access Membership)

A membership program belongs to a customer (store-level concept), whereas BTCPay Server users are a server-level concept.  

An administrator who wants to use a subscription to control access to BTCPay Server would configure it as follows:

* Go to **Settings**  
* Navigate to **Policies**  
* Set the currently selected store as the server membership (using the same UX as for the template store).

**TODO:** Decide how to handle mapping customers to existing accounts. Should we disable them? What if some accounts are administrators?

## Customer: Viewing History and Downloading Receipts

Customers should have access to a page where they can:

* View their payment history (similar to the Stripe payment page used by OpenAI).  
* Download receipts.  
* Extend their subscriptions.

## Customer: Renewal

Before the end of the billing period, an email should be sent to the customer with a link to the subscription page, prompting them to pay for the next period.

**Comment by pavlenex** [comment]
@NicolasDorier It sounds good to me. I am just concerned that we're again introducing a new concept - customer. We already have buyer in invoice, which is metadata, but overall these two are the same.

> The concept of a customer is useful in areas beyond subscriptions (for example, Pay Vendor plugin).
In the future, we could add the ability to associate a customer with an invoice. This would allow features like filtering invoices by customer.

Yes, this resolves my concern above, but let's actually commit and do it properly. We have a pattern where we always chase new shiny features and leave a lot of ambiguity on to-do that we never end up doing. So ACK only if we resolve the ambiguity between buyer metadata and customer as a concept and not end up having both.

**Comment by pavlenex** [comment]
Also I am unsure why membership is a store level concept, will you charge it per store or per account, per account may be more logical imo? Membership also just sounds like product concept. I would also look into how Stripe is managing these concepts in their API.

**Comment by robwoodgate** [comment]
As an ambassador thinking of opening up my instance, I also am concerned about:

- How to limit the number of stores a server user can create (eg set as part of their subscription plan)
- How I can terminate a store (if it is abusive or illegal etc). I can remove a user but not sure if this removes the store?
- How I can suspend a store from operating if their subscription lapses and reinstate it once the subscription is in order
- How to prevent subusers (non-owner) from subletting my instance resources unless this is part of their plan (ie Membership Program app is owner only, or is a subscription plan switch to allow resellers etc)

**Comment by NicolasDorier** [comment]
@pavlenex membership is a store level concept, because to pay a membership you need to generate an invoice, so you need a store.

By doing the way I am suggesting (membership at store level, then a settings in server settings to say "use this store for server access memberhsip"), the UX of subscription can be used in scenario outside purely monetizing the server without downside.

**Comment by NicolasDorier** [comment]
@robwoodgate 

> How to prevent subusers (non-owner) from subletting my instance resources unless this is part of their plan (ie Membership Program app is owner only, or is a subscription plan switch to allow resellers etc)

Can you detail more about this? You want to prevent customer A to charge customer B for using your server? I am not sure how we could prevent this?

**Comment by NicolasDorier** [comment]
>  Membership also just sounds like product concept.

But we don't have any product concept. (That said, I will look at stripe as well for inspiration)

> We already have buyer in invoice

The BuyerName in invoice isn't any concept at all, it is an arbitrary data we store in the metadata of invoice, a naming convention. There is no logic associated with it. Note that Stripe also has a Customer concept.

Ping @ndeet, in term of integration would that be an interesting concept?

**Comment by NicolasDorier** [comment]
@pavlenex  After a look at Stripe, here is what I noticed: They are using the concept of **Customer** and **Subscription** only.

The concept of `Subscription Plan` and `Membership` is outside the scope of stripe. The do have the concept of `Subscription Item` though. (For selling btcpay server access would be as @robwoodgate mentioned, the number of stores a server user can create)

I believe this is good idea. If we were to follow what they do it means that Customer and Subscriptions would become part of Core. Those are concepts relevant to payment processors, and it is a small scope.

However, the ability to define the different plans to access the server, which would rely on those subscriptions, would be left to a plugin that is only relevant to ambassadors.

**Comment by robwoodgate** [comment]
> [@robwoodgate](https://github.com/robwoodgate)
> 
> > How to prevent subusers (non-owner) from subletting my instance resources unless this is part of their plan (ie Membership Program app is owner only, or is a subscription plan switch to allow resellers etc)
> 
> Can you detail more about this? You want to prevent customer A to charge customer B for using your server? I am not sure how we could prevent this?

The primary concern is to ensure server resources are not abused and/or used up without being compensated or having any control, which in the case of BTCPay relates to the number of active stores on the server.

So if I sell a subscription to customer A, I would want to cap the number of stores they can create as part of that subscription.

In fact, regardless of whether I sell the access or just invite users, I would want to cap the number of stores each user can create - otherwise I could invite one user, and they could create an infinite number of stores... using up all my server resources!

I'm fine if they choose to resell/sublet stores _within their allocation_... as the overall number of stores they can have stays the same.

But I would want to ensure they cannot use the membership app/plugin to sell MORE stores than they are allocated.

### EXAMPLE (PROBLEM NOW):

I do not allow free registrations, but decide to invite a user to the server. Although they cannot create more USERS, they can abuse the server by creating an unlimited number of stores.

A bad actor could even sell those stores to others as a managed service, while I'm paying the bill for the server.

So, I would like to be able to set a store creation limit per user (with a server-wide default).

### EXAMPLE (WITH MEMBERSHIP APP):

Server Owner: Uses the membership app to create plans that grant access to the BTCPay server instance
and sets their store at the admin level to grant user access when a subscription it taken out.

Plan A: "Merchant - Monthly" - grants user access to create ONE store, billed monthly
Plan B: "Entrepreneur  - Monthly" - grants user access to create up to THREE stores, billed monthly
Plan C: "Reseller - Monthly" - grants user access to create up to TEN stores, billed monthly

Customers of these plans can all use the membership app in their store(s), as it is generally useful in contexts unrelated to granting access to a BTCPay Server instance.

They might even be able to sell memberships that sublet/resell stores within their allocation (I don't mind that).

But they cannot use the membership app to bypass their store creation limit.

Does that make sense?

**Comment by ndeet** [comment]
@NicolasDorier great work, as you already noticed I think we should heavily get inspiration from stripe, they thought through it a lot and it is very generic. Even if we have no product I would still name it like this because stripe is like the amazon S3 api (well not really but you get the idea, it is easier for integrators to understand as it is widely spread)

Asking Claude their data structure looks something like this

```
Product: (You named this Membership Program)
  name: "BTCPay Server Access"
  description: "Full access to BTCPay Server instance"
  
  Pricing Plans: (You named this Subscripion plan but then in the detailed desciption you put everything in the Membership Program)
    - name: "Monthly"
      interval: "month"
      interval_count: 1
      price: 10
      currency: "USD"
      
    - name: "Quarterly"
      interval: "month"
      interval_count: 3
      price: 27
      currency: "USD"
      
    - name: "Annual"
      interval: "year"
      interval_count: 1
      price: 100
      currency: "USD"

Subscription:
  customer_id: "cust_123"
  plan_id: "plan_monthly"
  status: "active"
  current_period_start: "2024-01-01"
  current_period_end: "2024-02-01"
```

This looks like a very generic and flexible structure, what you think?

**ad Customers**
we need to cover both cases, the customer in case of Ambassador is a user on the server but for external integrations they are bound to the store but the customer references are outside of BTCPay or they do not exist at all, so maybe external system has already a subscription system internally like [Woocommerce Subscriptions](https://woocommerce.com/products/woocommerce-subscriptions/) (note there is also a [WooCommerce Memberships](https://woocommerce.com/de/products/woocommerce-memberships/) plugin). The WooCommerce Subscriptions plugin basically just sends a unique subscription reference ID (it created on first purchase) and then tries to charge that again for the new period (change in prices also possible)

So I agree, we likely want a generic plugin that does only subscriptions/recurring payments and then a separate one that does specific BTCPay Ambassador/Renting stuff to not cram too much functionality in one plugin and keep it generic enough for all kinds of external systems that maybe already have their own membership system to integrate. We basically focus on recurring payments handling and sending mail reminders (if email available or just return a payment link for the new period) first, then add another plugin that does the server<->user<->store level membership management including permissions and resouce limitations.

**Comment by NicolasDorier** [comment]
@ndeet I think this is AI slop. The concept of subscription barely touch the concept of product (it can be ignored basically), and the concept of Pricing plan is completely separate from the concept of subscription.

---

Checking more they have two APIs. The one I talked about was the V1. This "Pricing plan" stuff is something separated they created recently V2. Both achieve the same goal, but IMHO the pricing plan is more integrated to the rest of the API while subscription is more contained.
So the first API is maybe better for us to restrict the scope.

I will also investigate https://woocommerce.github.io/subscriptions-rest-api-docs/#introduction

**Comment by NicolasDorier** [comment]
Woocommerce subscription only have a `Subscription` object, but this has a reference to `customer_id`. So it seems it relies only on Subscription and Customers as well.

**Comment by dstrukt** [comment]
These were my initial Figma explorations from a few months ago before any of the above discussion. 

I'm going to re-evaluate the mocks based on all of the above and post updated screens/flow. However, I still figured this would help push the discussion forward / give us something concrete to discuss. 

Apologies I didn't post this sooner.

# Goal

Give Server Admins the ability to charge for access to their instance. This is a basic prototype, that does not include all of the flow points in which the end user would be prompted to make a payment (upon Login, etc..), or be reminded of an upcoming payment (via email), but simply a starting point for functional discussion. 

These are still quite rough, with a lot of open questions, so this is not to be worked on until agreen upon.

https://github.com/user-attachments/assets/a26b2c60-deef-4bec-a879-d5ce4210aeab

# Three Tab Mocks (Recommended)

<img width="804" height="411" alt="Image" src="https://github.com/user-attachments/assets/f8b67815-d439-4aa8-b8b7-d96080e491e2" />
<img width="811" height="445" alt="Image" src="https://github.com/user-attachments/assets/4e57d53c-1bad-423b-98f8-451e7574c4f9" />
<img width="659" height="778" alt="Image" src="https://github.com/user-attachments/assets/7ef801eb-71ff-4b50-8bc8-7cf1a1ee4b51" />
<img width="757" height="860" alt="Image" src="https://github.com/user-attachments/assets/4b4a0e38-c991-4be8-857c-de44d0519357" />

# Single View Exploration (Not Recommended)

<img width="450" height="747" alt="Image" src="https://github.com/user-attachments/assets/0eb218ff-e68b-4cdc-ba70-28ce77ff3de4" />

**Comment by TChukwuleta** [comment]
> As an ambassador thinking of opening up my instance, I also am concerned about:
> 
> * How to limit the number of stores a server user can create (eg set as part of their subscription plan)
> * How I can terminate a store (if it is abusive or illegal etc). I can remove a user but not sure if this removes the store?
> * How I can suspend a store from operating if their subscription lapses and reinstate it once the subscription is in order
> * How to prevent subusers (non-owner) from subletting my instance resources unless this is part of their plan (ie Membership Program app is owner only, or is a subscription plan switch to allow resellers etc)



If we would consideer the number of stores when setting up membership, one thing to the consider is how to cater for upgrade and downgrade of membership.. 

say I want to switch from one membership to the other e.g. higher to lower.. 

which would then imply less stores...  with this, do we disable stores from admin level and if so what stores... 

pretty much managing conversation and handling use cases around a customer switching from one membership tier to another..

**Comment by NicolasDorier** [comment]
I investigated how stripe work, and I absolutely love it:

<img width="614" height="816" alt="Image" src="https://github.com/user-attachments/assets/3bf30bc7-c338-495b-9dcf-30bcc1391809" />

In the Stripe backend, the merchant sets up a membership with several Plans. A Plan can have multiple `Subscription Items` (called Products in Stripe’s terminology). Think of this as the maximum number of stores allowed for a customer upon a certain plan.

Each Plan has a shareable URL that can be copied and pasted into the merchant’s website call-to-action page (see the `Merchant Site` picture).

This URL sends the customer to a `subscription checkout page`, where they enter their email, credit card number, and other details. Upon successful payment, the customer is redirected to the merchant’s website, and webhooks are sent to the merchant.

The customer’s email creates a `Customer object` that can be viewed in the Stripe backend (this object is scoped to the store). You can search for and find customers, and clicking on one will show their payment history and a list of their subscriptions.

If the customer also has an account on the merchant’s website, they will see  `Manage subscription` button in their profile. Upon clicking it, they are redirected to stripe where they can view the current subscription status and choose to cancel or upgrade the subscription.

I find this UX super simple and easy to understand, and it is something we should have in BTCPay Server.

**Comment by NicolasDorier** [comment]
@dstrukt what is the difference between "Active users" and "Active subscriptions"?

Does it means that a user can have several subscription? This isn't something that exists in say Stripe afaik.

Your mock is showing "USD Price", but what about other currencies? Should there be a "Currency" field, or we take the default currency of the store? What if the store change the currency later?

The total revenue, in which currency it is? Is it the default currency of the store?
What if the user change currency? What if several plans are in different currency? How to add them together?

Maybe we simplify: We only count those equal to the same currency as the store? If yes, should we communicate this to the user?

-------

We could either go one store = one membership, but then I need proper answer on adding currencies and what represent any "total" we display.

Or we could go one store can have multiple membership, each membership having only a single currency, but then we need an additional screen to see the list of memberships.

-----

What are "Pending Payments" are they invoices that are processing but not settled?  In this case, why is it relevant data? this will be always between 0 and 2 or 3.
Or do you mean the members who still need to pay to get renewed?

**Comment by dstrukt** [comment]
All great questions, and as I'm revisiting these rough mocks with fresh eyes, I think they address a lot of our problem(s), but also lack quite some granular detail and finalized UI, i.e. any Store details, failure states and follow up Admin actions, etc.. Additionally, the mocks were solely focused on the Admin POV. 

I will also mock up the end-user A/B/C plan choice to help make the entire flow more clear.

> [@dstrukt](https://github.com/dstrukt) what is the difference between "Active users" and "Active subscriptions"?
> 
> Does it means that a user can have several subscription? This isn't something that exists in say Stripe afaik.

As I was thinking it through, I had considered the idea of a User being able to subscribe to multiple Stores, but also came to the conclusion that this is likely unnecessary, or perhaps more complicated than necessary for the MVP, and forgot to remove the `Active Users` section. I think in this case `Active Subscriptions` is a more relevant metric to keep.

> What are "Pending Payments" are they invoices that are processing but not settled? In this case, why is it relevant data? this will be always between 0 and 2 or 3. Or do you mean the members who still need to pay to get renewed?

Good question. Some of these aggregate metrics were just high level thinking of what *might* be useful - not 100% finalized in their current state.

I thought it might make sense to show both "Unpaid" (to be paid) and paid but "Unconfirmed" payments for a Subscription. Once paid, if the payments `fail`, the Admin or Customer would be notified that action was required. In both cases, the UI above is lacking a bit.

> Your mock is showing "USD Price", but what about other currencies? Should there be a "Currency" field, or we take the default currency of the store? What if the store change the currency later?
>
> The total revenue, in which currency it is? Is it the default currency of the store? What if the user change currency? What if several plans are in different currency? How to add them together?
> 
> Maybe we simplify: We only count those equal to the same currency as the store? If yes, should we communicate this to the user?
> 
> We could either go one store = one membership, but then I need proper answer on adding currencies and what represent any "total" we display.
>
> Or we could go one store can have multiple membership, each membership having only a single currency, but then we need an additional screen to see the list of memberships.

Yeah I just used USD as a consistent proxy for the mocks, but assumed we'd use the Store default, or perhaps a single denomination in BTC? But I wasn't 100% sure the technical limitations we would face, or how complicated it would ultimately be / assumed we'd discuss this in a call! 

We could also scrap the idea of a `Total` if it's too complicated ultimately, but I just figured it would be nice for a Server/Store admin to see the total revenue generated from their subscribed users.

**Comment by NicolasDorier** [comment]
> Yeah I just used USD as a consistent proxy for the mocks, but assumed we'd use the Store default, 

This is not a technical limitation, but a UX one. A store can have a default currency, but the default currency can be changed by the user at any time.

If you use the store's currency for calculating the total, then we can't show the total in the new currency anymore once the merchant change the default currency, because the existing plans were paid in a currency different from the store's one.
This only way to fix this issue is to make it impossible to change the currency of a membership program.

But then this mean we need an additional step, where the merchant create a membership program. (I guess only name and currency are really needed as settings)
And we need an additional page to browse them. (Or maybe just a drop down in the memberships page?)

**Comment by NicolasDorier** [comment]
So I looked to how stripe are doing things.
Stripe has the concept of `Pricing`. You can affect several pricing (different amount/currency) to a single plan.
There seem to be no interface summing up everything...

Maybe in our case it's a bit too much and having a subscription program with fixed currency is better?

**Comment by dstrukt** [comment]
Ahh alright, well thinking through this.

I'd imagine we wouldn't make a Subscription plan editable once it's been created (to prevent Admin abuse - switching the cost amount, frequency, or currency on subscribers, etc..), they can add a new plan to the list of plans, but not modify an existing. 

Which then leads me to question how a Merchant and Customer would realistically want to use this and how they'd expect to pay and be paid. 

In most cases, they'd probably expect a consistent currency across the Subscription plans / tiers, i.e. USD, BTC, EUR, etc.. but I can imagine Merchants and Customers *might* want the flexibility of different payment methods too, which leads me to think having a flexible choice in price/currency for each plan is worth foregoing the aggregated sum total of revenue for a Membership Plan or of all Membership plans on the Store. That's more of a vanity metric, and not necessarily something we should prioritize if it's difficult to sum. We could make the first dropdown choice the Store default currency, but allow the Merchant to select something different if they choose).

If we wanted, we could total up how much revenue each specific Subscription plan has generated (i.e. Plan A generated $80 USD, Plan B generated $0.32 BTC, Plan C generated 30 EUR), but maybe not sum all of the Membership Subscription into a singular total.

I do like the simplicity of a fixed currency for a Membership and it's child Subscription plans (makes sums and totals much easier), but I fear it will ultimately be too limiting.

**Comment by dstrukt** [comment]
I'm also still in favor of calling this plugin `Subscriptions` instead of `Membership`, as you create a `Subscription`, thus have `Subscriptions` with various sub `Plans` (30/mo, 100/yr, etc..).

I think that wording is more intuitive, and terms generally more well-known to the public as well.

**Comment by TChukwuleta** [comment]
> So I looked to how stripe are doing things. Stripe has the concept of `Pricing`. You can affect several pricing (different amount/currency) to a single plan. There seem to be no interface summing up everything...
> 
> Maybe in our case it's a bit too much and having a subscription program with fixed currency is better?

Yes, this sounds okay..  Fix currency immediately after first subscriber. THe merchant can decide to clone the current plan (say v2 or plan - pro) if changes are needed. 


Regarding this @dstrukt :  "but I fear it will ultimately be too limiting."

I believe the perceived limitation is actually all the features that makes up that subscription plan.. I think that can then be modifiable to allow more features or remove from it

**Comment by robwoodgate** [comment]
I think terminology is the biggest hurdle when defining these things... here's how aMember (a PHP membership script that I wrote [a BTCPay Plugin](https://github.com/robwoodgate/am-btcpay) for) broadly organises entities:

- **Products** - This describes the offering in general terms - eg: title, description, category etc
- **Billing Plans** - The specific subscription parameters for a Product - eg: currency, first price, first period, rebill times (never, once, x times, until cancelled), second price, second period. A Product can have multiple Billing Plans.
- **Customer** -  The purchaser of a Product - eg: name, email etc
- **Invoice** - Records the specific Billing Plan chosen by customer and the subscription status - eg: active, cancelled
- **Payment** - Records the subscription payment(s) made against the invoice and creates Access records
- **Access** - Records the paid up period of the product subscription and the Product / Customer / Invoice / Payment it relates to.

A subscription is cancelled by marking the invoice subscription status as cancelled. This stops Payments.
When paid up Access expires, the permissions granted by the Product via that Invoice are revoked.

For BTCPay ambassadors selling stores, a Product could grant access to create X number of stores. 
Handling subscription expiry is a matter of personal preference (or convenience for you guys)...

My preference would be for the customer to have a "store quota", being the sum of all the active subscriptions granting stores, and then have some UI to reallocate that quota to stores - like a "license". 

So for example, If a customer had one subscription to a product granting 2 stores, and another subscription to a product granting 1 store, they have a 3 store quota. They have 3 active stores, so the whole quota is used. They decide to cancel the subscription to the 1 store product, and when the paid up Access period expires, they will only have quota for 2 stores.
They can either choose to deactivate the "licence" on the store they want to close before the access expires - in which case, they will still be in quota when the subscription expires... or the server will deactivate one for them at expiry (maybe most recently created or at random). In this situation, the customer can login and deactivate the license on the store they intended to close, and add it to the store that was deactivated to reactivate it.
 

> I'd imagine we wouldn't make a Subscription plan editable once it's been created (to prevent Admin abuse - switching the cost amount, frequency, or currency on subscribers, etc..), they can add a new plan to the list of plans, but not modify an existing.

In the above model, once a purchase is made, the billing plan chosen is baked into the invoice, so plans can be changed later and will not affect existing subscribers.

**Comment by NicolasDorier** [comment]
> Merchants and Customers might want the flexibility of different payment methods too, which leads me to think having a flexible choice in price/currency for each plan is worth foregoing the aggregated sum total of revenue for a Membership Plan or of all Membership plans on the Store. 

The currencies of payment method don't have the be the same as the currency of the invoices being paid. Typically, invoices would be USD despite the payment being done in BTC or any other crypto. Then the UI would only show the USD amount.

The concept of "pricing" in Stripe is to say: Ok, this plan cost 1,000 JPY for japanese, and 10 USD for US.

> I do like the simplicity of a fixed currency for a Membership and it's child Subscription plans (makes sums and totals much easier), but I fear it will ultimately be too limiting.

Yeah, it might be.

> Fix currency immediately after first subscriber. The merchant can decide to clone the current plan (say v2 or plan - pro) if changes are needed.

Yeah maybe that is enough... so the currency would be fixed at the plan level

> I think terminology is the biggest hurdle when defining these things..

I think I made it good on this, I'd like to decrease the number of concepts. What I am thinking right now is:

* Customers: A concept independent from subscriptions, might use it for other things in BTCPay (for example for storing emails, nostr, add logs specific to customer, providing an easy way to see latest invoices or subscriptions etc...)
* Members: A concept for subscription, it links to a customer, and hold subscription data (the state, the end of the current period, the current plan)
* Plans: Currency / Price / Period and also "Subscription items" (in our example it could be `3 stores` if you want to sell access to 3 store )

I think this would cover all the use cases.
I am wondering to add the concept of `Membership program` to tie several plans together. So one store can be used for more than 2 subscription services... but unsure if it is needed. For now plans are directly at the store level.

**Comment by NicolasDorier** [comment]
@dstrukt other kind of things that should be possible in the backend:

* Ability to force deactivate a user, despite being in a paid period
* How we handle plan migration? (In stripe, it seems you can upgrade, and the prorated part of the current period already used count to decrease the cost of the upgrade)
* Ability to give a free trial

**Comment by NicolasDorier** [comment]
@dstrukt other things to think about emails:

* Encourage upgrade before trial ends.
* Nudge the user to renew the plan as he enters is in the Grace period
* Nudge the user to pay for the next month while still in the current period
* Email to the user when the grace period ends
* Capture lost users by sending email after they stop the trial period but didn't pay

How can we present those settings and make them configurable? We have the email rules, so I guess any UX will just create rule their.

**Comment by dstrukt** [comment]
> I think this would cover all the use cases. I am wondering to add the concept of `Membership program` to tie several plans together. So one store can be used for more than 2 subscription services... but unsure if it is needed. For now plans are directly at the store level.

I think that's a follow up feature or at least wait until it's user request, imo.

**Comment by dstrukt** [comment]
> * Ability to force deactivate a user, despite being in a paid period

Hmm, I'd think if the Admin has the option to disable that Customer, then it's just be a gated conditional when the Customer tries to log in, no? We should potentially flash an error when the deactivated/banned user attempts to log in `Your account has been deactivated. Contact the Store Admin.` sort of thing?

> * How we handle plan migration? (In stripe, it seems you can upgrade, and the prorated part of the current period already used count to decrease the cost of the upgrade)

Hmm, good question. We could potentially omit this for V1 and let Admins/Customers sort it out, otherwise, yeah we'll need to be to calculate the pro-rated amount to inform the Store Admin for either a refund or to calculate against the updated price for that pay period to show to the Customer. The latter is arguably a cleaner UX, but more complicated for us to implement - would be sorted out in the Customer Portal though.

> * Ability to give a free trial

Agreed, and I think each Plan will have a `Trial Period (Days)`, and if > 0 this would allow the Customer access until it's over, then they're hit with either the Plan decision view, or an invoice, if we make them select a Plan before starting the Trial. We could display the number of Trial days left somewhere to Customer, otherwise it's just a calculation of when the Billing/Usage period has started, then the Trial ended message, time to pay.

**Comment by dstrukt** [comment]
> [@dstrukt](https://github.com/dstrukt) other things to think about emails:
> 
> * Encourage upgrade before trial ends.
> * Nudge the user to renew the plan as he enters is in the Grace period
> * Nudge the user to pay for the next month while still in the current period
> * Email to the user when the grace period ends
> * Capture lost users by sending email after they stop the trial period but didn't pay
> 
> How can we present those settings and make them configurable? We have the email rules, so I guess any UX will just create rule their.

I think I have a few of these cases mocked up in the Settings tab above. Basically a generated/editable email template that goes out to the Customer based on the conditionals. I think we do something similar in the Payroll plugin IIRC, cc @rockstardev.

**Comment by dstrukt** [comment]
Sorry for any delay, trying to work through this ASAP so you can keep moving.

# Customer Portal

We could accommodate multiple subscriptions to different Stores on the same Server, if we choose to support that. Need to work through the Renew & Change Plan modals though, TBD.

<img width="975" height="492" alt="Image" src="https://github.com/user-attachments/assets/6b8f06a7-8874-4a62-af91-83927d8481fe" />

# Store Admin Portal

Still a bit more to do on the finalization of the UI, but don't want to slow us down, so will just keep posting progress shots and videos.

https://github.com/user-attachments/assets/1a34767c-8d84-4d50-a4e1-6e5715084f2a

# Subscribers/Customer Tab (Store Admin Portal)

We have two choices here, break Customers/Subscribers into groups based on their Membership, or just dump them all out in a singular table with the correlating Membership in the column. I think I'm leaning towards the latter, makes search/filtering/sorting/etc.. a bit easier, but thought I'd share in case you prefer the former.

<img width="738" height="444" alt="Image" src="https://github.com/user-attachments/assets/fab85615-e5f5-4494-be38-d781ea0cc196" />
<img width="746" height="397" alt="Image" src="https://github.com/user-attachments/assets/df817cff-47d6-4d1e-b9e4-c6302a85c770" />

# Choose Plan (Customer Flow)

You posted a rough idea of this, but TBD, basically present the plan choice -> input email -> generate invoice -> present & email customer portal link -> allow access.

Excited to see initial PRs of all this.

**Comment by NicolasDorier** [comment]
@dstrukt there is also 4 phases, which I think could be good to show.

* In Trial
* Normal
* Grace
* Expired

This is orthogonal to `Active/Disabled`. Since a member could be disabled even if he is in the Normal phase by the admin.

> We should potentially flash an error when the deactivated/banned user attempts to log in

So the membership is separate from the monetization of the server. 
I will later work on the monetization of the server, which will use membership.

In membership what happen is that when a subscriber is Actived or Disabled, a webhook is sent. This webhook is used by the integrator to enable/disable a service. The UI of such service is out of the scope of BTCPay.
The server monetization is such a service, but this is out of the scope at this point.

**Comment by NicolasDorier** [comment]
@dstrukt there is no "Add a plan" button.

Maybe another idea is also to follow what we do for PoS app: We put the merchant's "Subscriptions" below the "Subscriptions" section in the sidebar. Then he can click on the one he is interested in.

**Comment by dstrukt** [comment]
> [@dstrukt](https://github.com/dstrukt) there is no "Add a plan" button.
> 
> Maybe another idea is also to follow what we do for PoS app: We put the merchant's "Subscriptions" below the "Subscriptions" section in the sidebar. Then he can click on the one he is interested in.

**I re-uploaded the prototype video file in my previous comment, I don't think it uploaded correctly.

Great catch, yes, Add Plan should also be up there as well!

Interesting, I like the sidebar idea, esp. if more detail is added to the them later. Can always for loop stack them in the parent too, but easy enough to prototype both and see what we like!

**Comment by dstrukt** [comment]
> [@dstrukt](https://github.com/dstrukt) there is also 4 phases, which I think could be good to show.
> 
> * In Trial
> * Normal
> * Grace
> * Expired
> 
> This is orthogonal to `Active/Disabled`. Since a member could be disabled even if he is in the Normal phase by the admin.

Sounds good, we can just use the Invoice tag components and rename them to the phases/states we want.
 
> > We should potentially flash an error when the deactivated/banned user attempts to log in

Agreed, good to communicate this.
 
> So the membership is separate from the monetization of the server. I will later work on the monetization of the server, which will use membership.

Yeah, this is an important one, but a bit tricky with restrictions, all the considerations mentioned above, etc.. but we could potentially make it a dedicated / special Membership, with permission conditions, more clearly labelled functionality, etc.. curious to hear your thoughts on this as we keep building.

> In membership what happen is that when a subscriber is Actived or Disabled, a webhook is sent. This webhook is used by the integrator to enable/disable a service. The UI of such service is out of the scope of BTCPay. The server monetization is such a service, but this is out of the scope at this point.

Thanks for the clarification - good to know.

**Comment by NicolasDorier** [comment]
WIP https://github.com/btcpayserver/btcpayserver/pull/6922
